src/adaptix/_internal/model_tools/introspection/sqlalchemy.py

```diff
@@ -0,0 +1,256 @@
+    DefaultValue,
+    FullShape,
+    InputField,
+    InputShape,
+    IntrospectionImpossible,
+    NoDefault,
+    NoTargetPackage,
+    OutputField,
+    OutputShape,
+    PackageIsTooOld,
+    Param,
+    ParamKind,
+    Shape,
+    create_attr_accessor,
+)
+
+T = TypeVar('T')
+
+
+class IdWrapper(Generic[T]):
+    def __init__(self, value):
+        self.value = value
+
+    def __eq__(self, other: Any):
+        if isinstance(other, IdWrapper):
+            return self.value is other.value
+        return NotImplemented
+
+    def __repr__(self):
+        return f'{type(self).__name__}({self.value})'
+
+
+def _is_context_sensitive(default: 'CallableColumnDefault'):
+    try:
+        wrapped_callable = default.arg.__wrapped__  # type: ignore[attr-defined]
+    except AttributeError:
+        return True
+
+    parameters = inspect.signature(wrapped_callable).parameters
+    return len(parameters) > 0
+
+
+def _unwrap_mapped_annotation(type_hint: TypeHint) -> TypeHint:
+    if get_origin(type_hint) == Mapped:
+        return get_args(type_hint)[0]
+    return type_hint
+
+
+def _get_type_for_column(column: 'ColumnElement', type_hints: Mapping[str, TypeHint]):
+    try:
+        return _unwrap_mapped_annotation(type_hints[column.name])
+    except KeyError:
+        if column.nullable:
+            return Optional[column.type.python_type]
+        return column.type.python_type
+
+
+def _get_type_for_relationship(relationship: 'RelationshipProperty', type_hints: Mapping[str, TypeHint]):
+    try:
+        return _unwrap_mapped_annotation(type_hints[relationship.key])
+    except KeyError:
+        if relationship.uselist:
+            return List[relationship.entity.class_]  # type: ignore[name-defined]
+        return Optional[relationship.entity.class_]
+
+
+def _get_default(column_default: 'Optional[DefaultGenerator]'):
+    if isinstance(column_default, CallableColumnDefault):
+        if _is_context_sensitive(column_default):
+            return NoDefault()
+        return DefaultFactory(factory=column_default.arg.__wrapped__)  # type: ignore[attr-defined]
+    if isinstance(column_default, ScalarElementColumnDefault):
+        return DefaultValue(value=column_default.arg)
+    return NoDefault()
+
+
+def _is_input_required_for_column(column: 'ColumnElement', autoincrement_column: 'Optional[Column[int]]'):
+    return not (
+        # columns constrained by FK are not required since they can be specified by instances
+        column.default is not None
+        or column.nullable
+        or column.server_default is not None
+        or column.foreign_keys
+        or column is autoincrement_column
+    )
+
+
+def _get_autoincrement_column(table: 'sqlalchemy.Table'):
+    try:
+        return table.autoincrement_column
+    except AttributeError:
+        # for old sqlalchemy
+        return table._autoincrement_column  # pylint: disable=protected-access
+
+
+def _get_input_shape(
+    tp: TypeHint,
+    table: 'FromClause',
```

Comment:
table has concrete type `sqlalchemy.Table` because of instance checking [here](https://github.com/reagento/adaptix/pull/240/files#diff-5ec167b8b99d007e1b922e15e3b05d36ac975a7f4c8b209f29dca92426d1c663R237-R240)